library/src/compute/construct.rs

```diff
@@ -184,6 +186,185 @@ cast_from_value! {
+                Ok(date) => Some(date),
+                Err(_) => bail!(args.span, "date is invalid"),
+            }
+        }
+        _ => None,
+    };
+
+    match (date, time) {
+        (Some(date), Some(time)) => Value::Dyn(Dynamic::new(Datetime::Datetime(
+            PrimitiveDateTime::new(date, time),
+        ))),
+        (Some(date), None) => Value::Dyn(Dynamic::new(Datetime::Date(date))),
+        (None, Some(time)) => Value::Dyn(Dynamic::new(Datetime::Time(time))),
+        (None, None) => {
+            bail!(args.span, "at least one of date or time must be fully specified")
+        }
+    }
+}
+
+struct YearComponent(i32);
+
+struct MonthComponent(Month);
+
+struct DayComponent(u8);
+
+struct HourComponent(u8);
+
+struct MinuteComponent(u8);
+
+struct SecondComponent(u8);
+
+cast_from_value!(
+    YearComponent,
+    v: i64 => match i32::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("year is invalid")?
+    }
+);
+
+cast_from_value!(
+    MonthComponent,
+    v: i64 => match u8::try_from(v) {
+        Ok(n1) => match Month::try_from(n1) {
+            Ok(n2) => Self(n2),
+            _ => Err("month is invalid")?
+        }
+        _ => Err("month is invalid")?
+    }
+);
+
+cast_from_value!(
+    DayComponent,
+    v: i64 => match u8::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("day is invalid")?
+    }
+);
+
+cast_from_value!(
+    HourComponent,
+    v: i64 => match u8::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("hour is invalid")?
+    }
+);
+
+cast_from_value!(
+    MinuteComponent,
+    v: i64 => match u8::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("minute is invalid")?
+    }
+);
+
+cast_from_value!(
+    SecondComponent,
+    v: i64 => match u8::try_from(v) {
+        Ok(n) => Self(n),
+        _ => Err("second is invalid")?
+    }
+);
+
+/// Returns the current date.
+///
+/// If no offset is specified, the date according to the local time will be chosen.
+/// If an offset is specified, the current UTC date with the applied offset will be chosen.
+///
+/// ## Example
+/// ```example
+/// Today's date is #today().display("[month repr:long] [day], [year]")
+/// ```
+///
+/// Display: Today
+/// Category: construct
+/// Returns: datetime
+#[func]
+pub fn today(
+    /// Whether an offset should be applied. If none is specified, the local date will be chosen.
+    /// If an offset is specified, it will be applied to the current UTC date.
+    #[named]
+    offset: Option<i64>,
+) -> Value {
+    let current_date = match vm.vt.world.today(offset) {
+        Some(d) => d,
+        None => bail!(args.span, "unable to get the current date"),
+    };
+
+    match time::Date::from_calendar_date(
+        current_date.0,
+        time::Month::try_from(current_date.1).unwrap(),
+        current_date.2,
+    ) {
+        Ok(d) => Value::Dyn(Dynamic::new(Datetime::Date(d))),
+        Err(_) => bail!(args.span, "unable to read the system date"),
```

Comment:
If, for some reason, the `today` function returns a triple that does not constitue a valid date, e.g. (2023, 4, 32). This obviously should in theory never happen. Should I just unwrap instead?

Reply:
Perhaps replacing the error message with something like `"system returned an invalid date"` would be clearer then

Reply:
Yeah I like this one! Will change it.

Reply:
Couldn't the World's `today` function just return a `Datetime` instead? Then, it can't be invalid.

Reply:
It could, but then everyone would be forced to use the `time` crate when implementing the `World` trait (e.g. we would also need it as a dependency in `tests` and `docs`), I thought it would be better to make it crate-agnostic by just returning a tuple and leaving it up to the implementation to decide how the date is calculated. But if you think it's worth the tradeoff, I can change it.

Reply:
Maybe that's an indicator that the `Datetime` type is missing a few simple Rust constructors?

Reply:
Fair enough, will change this. :)

Reply:
I hope it's good now.